[Dependency Scanner] Batch scanner: extract target triple from the PCMArgs for the batch invocation.

artemcm opened the pull request:
Otherwise the created invocation will end up re-using the parent compiler invocation's target triple for every created scanning instance.

I stumbled upon this when testing https://github.com/apple/swift-driver/pull/225. 
The PCMArgs we pass in to the batch scanner are intended to be clang arguments and are prefixed with `-Xcc`, so they do not affect the created compiler instance.

artemcm commented:
@swift-ci please smoke test

nkcsgexi commented on lib/FrontendTool/ScanDependencies.cpp:
```diff
@@ -719,6 +719,18 @@ bool swift::batchScanModuleDependencies(CompilerInvocation &invok,
+            subInvok.setTargetTriple(*(it+2));
+          }
+        }
+      }
```
Should we extract all these `-xcc` arguments to add to the compiler instance via `setExtraClangArgs`?

artemcm replied:
Oh, for some reason I assumed that this would be taken care of when we pass them into `CompilerInvocation::parseArgs(`. But if it is not, then yes, we absolutely should.

nkcsgexi replied:
Even better, we need to teach `parseArgs` to parse these extra clang arguments properly.

artemcm replied:
After teaching `parseArgs` to set up `ClangImporterOpts.ExtraArgs` we would still require this change though. Is it alright to do it in a followup?

nkcsgexi replied:
Could you clarify why we still need this change?

nkcsgexi replied:
It seems to me that `parseArgs()` is handling `-Xcc` arguments properly. The real issue here is when we are creating a clang instance for dependencies scanning, we never use these extra arguments (see ClangImporter::getModuleDependencies for details).

artemcm replied:
My understanding is that even with `ClangImporterOpts.ExtraArgs` set, the target that gets used on the clang `scan-deps` invocation comes from the `ASTContext` on which we will call `getModuleDependencies`. 
That context belongs to the instance we set up here using the invocation on which we call `parseArgs`. Without this change (extracting the `-target`), this context will inherit the parent invocation's target, regardless of whether or not we correctly parsed `ExtraArgs`.

artemcm replied:
> It seems to me that `parseArgs()` is handling `-Xcc` arguments properly. The real issue here is when we are creating a clang instance for dependencies scanning, we never use these extra arguments (see ClangImporter::getModuleDependencies for details).

Oh, yes. I think we should then make sure we go ahead and use them in `getClangDepScanningInvocationArguments`, then.

nkcsgexi replied:
Yeah, that sounds to me the correct solution.

artemcm replied:
It was a bit more plumbing than I'd anticipated, but it is now ready for a review.

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please smoke test

nkcsgexi commented on lib/ClangImporter/ClangModuleDependencyScanner.cpp:
```diff
@@ -348,18 +353,14 @@ bool ClangImporter::addBridgingHeaderDependencies(
-  // Reform the Clang importer options.
-  // FIXME: Just save a reference or copy so we can get this back.
-  ClangImporterOptions importerOpts;
-
```
hmm, is this plumbing necessary if we add here `importerOpts.ExtraArgs = getExtraClangArgs();`?

artemcm replied:
🤦‍♂️ Can't believe I missed that. 
Good catch, thank you!

artemcm committed:
[Dependency Scanner] Ensure that Clang dependency scanner instances inherit the creating invocation's extra clang args.
This ensures that when the dependency scanner is invoked with additional clang (`-Xcc`) options, the Clang scanner is correctly configured using these options.

artemcm commented:
@swift-ci please smoke test

nkcsgexi reviewed (approved):
Thank you!👍

artemcm commented:
@swift-ci please smoke test OS X platform
